Initial marzipan development

meiertgrootes opened the pull request:
Hi @Francesco. have a look what you think of this. Dockerization and Emma to follow shortly

meiertgrootes committed:
initial commit of marzipan

meiertgrootes committed:
update README and added doc strings and requirements.txt

fnattino commented on marzipan.py:
```diff
@@ -0,0 +1,128 @@
+                inprogress = 0
+                print('All cluster VMs in state RUNNING')
+            else:
+                wait(10)
```
```suggestion
                time.sleep(10)
```

fnattino commented on marzipan.py:
```diff
@@ -0,0 +1,128 @@
+        pwd = self.config['one']['password']
+        endpoint = self.config['one']['endpoint']
+        one_server = pyone.OneServer(endpoint, session = uname+':'+pwd)
+        self.client = one_server
```
It is true that this is the client to connect to the open nebula end point, but maybe it's confusing to call client the instance of the server class?

meiertgrootes replied:
@fnattino yes, good point. how would you feel about `one_client` ?

fnattino replied:
sounds good, I think it would be much clearer!

fnattino reviewed (commented):
I have run it myself and I managed to create VMs very smoothly - this is sweet indeed ;) 
I just have only two specific comments (see code). We might think about adding additional methods to 'clean up' the provisioned clusters (delete VM instances and templates) and a method to write the hostfile to be used by Ansible, but for now looks great!

meiertgrootes committed:
Update marzipan.py

Co-authored-by: Francesco Nattino <49899980+fnattino@users.noreply.github.com>

meiertgrootes committed:
full intial version of docker framework. provisioning not yet implemented`

meiertgrootes committed:
fixed merge conflict

meiertgrootes commented:
Hi @fnattino,
I've added the full Docker framework, including the construction of the hosts.yaml file which is written back out to the user. I've also updated the readme accordingly. Please have a look. If you're ok with it I will merge to `main`

meiertgrootes committed:
Update README.md

meiertgrootes committed:
added ansible provisioning. Still working on platform install

meiertgrootes commented:
Hi @fnattino I've added in some of the Ansible provisioning. Basically the initial set up part. The platform installation still gives rise to a yaml syntax error (or at least the trae says so, although that may be misleading apparently). This will require some fiddling with Emma, but should hopefully work out fairly quickly.

meiertgrootes committed:
adapted Dockerfile to use emma_marzipan emma branch

meiertgrootes committed:
modified inventory builder to start indexiing at 1. TODO: check for issues wih adding hosts

meiertgrootes committed:
updated marzipan deploy to work with modified emma

meiertgrootes commented:
Hi @fnattino, I've added full Emma functionality including the prep-cloud and install_platform_light playbooks, he latter with common and dask roles. It's now ready for review and merging as fully functional provisioner.

fnattino commented on marzipan_scripts/marzipan_deploy.py:
```diff
@@ -0,0 +1,92 @@
+#construct command
+deploymentFolder = surfone.config["cluster"]["basename"]+"-cluster"
+deploymentFolderPath = "/marzipan/deployments/"+deploymentFolder
+rc = subprocess.run(["mv","/marzipan/deployments/tmp",deploymentFolderPath])
```
If a cluster has already been deployed, and one tries to deploy a new cluster with the same name, the `tmp` folder with the keys ends up as a subdirectory of `deploymentFolderPath`, making the provisioning fail in the inventory builder (not sure why). Maybe one should check whether a cluster with the same name has already beed deployed (i.e. if the folder exists in deployments) and raise an error if this is the case?

meiertgrootes replied:
this is strange, and I can't reproduce your findings. What I ge when trying to deploy an existing cluster is also a failure (which is desirable in my opinion), but it fails at the template constructon step, as the template name is already taken, so basically as below. at that point, the keys that have been generated are located in `/marzipan/deployments/tmp` (at least for me).

Let's briefly discuss, and then we can decide how/whether to catch this case.

```
Generating the ssh-key for root user.
Generating the ssh-key for ubuntu user.
initializing interface ...
reading configuration file ...
retrieving publuc key ...
reading template ...
connecting to ONE server ...
creating template on ONE ...
Traceback (most recent call last):
  File "/marzipan/marzipan_deploy.py", line 13, in <module>
    surfone = deploy_cluster()
  File "/marzipan/Marzipan/marzipan.py", line 142, in deploy_cluster
    surfone.create_one_template()
  File "/marzipan/Marzipan/marzipan.py", line 58, in create_one_template
    self.client.template.allocate(self.template)
  File "/usr/lib/python3.6/xmlrpc/client.py", line 1112, in __call__
    return self.__send(self.__name, args)
  File "/usr/local/lib/python3.6/dist-packages/pyone/__init__.py", line 236, in _ServerProxy__request
    return self.__response(ret)
  File "/usr/local/lib/python3.6/dist-packages/pyone/__init__.py", line 286, in __response
    raise OneInternalException(message)
pyone.OneInternalException: [TemplateAllocate] NAME is already taken by TEMPLATE 15552.
```

meiertgrootes replied:
I've reproduced the failure you identified. I've adapted marzipan and the deployment script to handle the situation. see below.

fnattino commented on config/ClusterConf.ini.example:
```diff
@@ -0,0 +1,14 @@
+endpoint = https://api.hpccloud.surfsara.nl/RPC2
+
+[ssh]
+root_public_key = /Users/eslt0101/Projects/REPOS/eScience/GO/marzipan/id_rsa_lokum_root.key.pub
```
The root public key is generated by Marzipan (with the `marzipan_scripts/generate_keys.sh`), right? Can we pick the path from there and remove this field here?

meiertgrootes replied:
Actually this is what is happening. I had forgotten to update the example .ini file. the actual one isn't shared with GitHub in order to protect the user password. I've updated the example now. This should also deal with the issue below.

fnattino commented on config/ClusterConf.ini.example:
```diff
@@ -0,0 +1,14 @@
+root_public_key = /Users/eslt0101/Projects/REPOS/eScience/GO/marzipan/id_rsa_lokum_root.key.pub
+
+[cluster]
+VMtemplate = /Users/eslt0101/Projects/REPOS/eScience/GO/marzipan/opennebula_goera.tpl
```
Since the deployment runs within the docker container, this path need to be specified as it is mounted in the docker image file-system (`/marzipan/...`), right? It's a bit tricky to 'guess' the path, I wonder whether we should just assume the template file should have a fixed name and be in a fixed place..

meiertgrootes replied:
see response to previous comment

fnattino reviewed (changes requested):
Hi @meiertgrootes, sorry for taking so long with this.. This is really great, it works very nicely and the instructions are clear. I have left only a couple of comments, but this is super promising!

meiertgrootes committed:
updated config example. set install_platform to run with extra_python_packages tag

meiertgrootes committed:
added error handliing for existing template

meiertgrootes committed:
added functionlity to catch edge case of instantiation of previously existing cluster whocch has been removed from ONE, but files remain in deployments

meiertgrootes commented:
hi @fnattino I've introduced code to check for the existence of the `deploymentPathFolder`. I've left this where it was in the deployment script, but added methods to terminate VMs and clusters, as well as to delete templates to marzipan.
 This should deal with all the issues you raised. please check, and then we can hopefully merge de elopement to main.

fnattino commented:
Looks great, thank you!

fnattino reviewed (approved):
Again, very nice :) !
